File: Day-167(DSA-practice)/app1.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
using namespace std;

string firstNonRepeating(string s) {
    unordered_map<char, int> freq;
    queue<char> q;
    string result = "";

    for (char ch : s) {
        freq[ch]++;
        q.push(ch);

        while (!q.empty() && freq[q.front()] > 1) {
            q.pop();
        }

        if (q.empty())
            result += '#';
        else
            result += q.front();
    }
    return result;
}

int main() {
    string s = "aabc";
    cout << firstNonRepeating(s);
    return 0;
}

// Output:
// a#bb

File: Day-167(DSA-practice)/app1.py
from collections import deque

#using queue
def first_non_repeating(s):
    freq = {}
    q = deque()
    result = ""

    for ch in s:
        freq[ch] = freq.get(ch, 0) + 1
        q.append(ch)

        # Remove characters from front that are repeating
        while q and freq[q[0]] > 1:
            q.popleft()

        # Append result
        if q:
            result += q[0]
        else:
            result += '#'

    return result

# Example
s = "aabc"
print(first_non_repeating(s))   # Output: a#bb
